compiler/modules/nback/build.bal

```diff
@@ -916,41 +916,171 @@ function buildBitwiseBinary(llvm:Builder builder, Scaffold scaffold, bir:IntBitw
+    llvm:Value rhsIsNill = builder.iCmp("eq", rhs, llvm:constNull(llvm:pointerType("i8", 1)));
+    llvm:Value eitherNill = builder.iBitwise("or", lhsIsNill, rhsIsNill);
+
+    llvm:BasicBlock eitherNillBB = scaffold.addBasicBlock();
+    llvm:BasicBlock neitherNillBB = scaffold.addBasicBlock();
+    llvm:BasicBlock joinBB = scaffold.addBasicBlock();
+    builder.condBr(eitherNill, eitherNillBB, neitherNillBB);
+
+    builder.positionAtEnd(eitherNillBB);
+    llvm:Value bothNill = builder.iBitwise("and", lhsIsNill, rhsIsNill);
+
+    if insn.op is "<=" || insn.op is ">=" {
+        buildStoreBoolean(builder, scaffold, bothNill, insn.result);
+    }
+    else {
+        buildStoreBoolean(builder, scaffold, llvm:constInt(LLVM_BOOLEAN,0), insn.result);
+    }
+    builder.br(joinBB);
+
+    builder.positionAtEnd(neitherNillBB);
+    bir:OptOrderType orderTy = <bir:OptOrderType> insn.orderType;
+    match orderTy.opt {
         t:UT_INT => {
-            buildStoreBoolean(builder, scaffold,
-                              builder.iCmp(buildIntCompareOp(insn.op),
-                                           buildInt(builder, scaffold, <bir:IntOperand>insn.operands[0]),
-                                           buildInt(builder, scaffold, <bir:IntOperand>insn.operands[1])),
-                              insn.result); 
+            llvm:Value lhsUntagged = buildUntagInt(builder, scaffold, <llvm:PointerValue>lhs);
+            llvm:Value rhsUntagged = buildUntagInt(builder, scaffold, <llvm:PointerValue>rhs);
+            buildCompareInt(builder, scaffold, buildIntCompareOp(insn.op), lhsUntagged, rhsUntagged, result);
         }
         t:UT_FLOAT => {
-            buildStoreBoolean(builder, scaffold,
-                              builder.fCmp(buildFloatCompareOp(insn.op),
-                                           buildFloat(builder, scaffold, <bir:FloatOperand>insn.operands[0]),
-                                           buildFloat(builder, scaffold, <bir:FloatOperand>insn.operands[1])),
-                              insn.result); 
+            llvm:Value lhsUntagged = buildUntagFloat(builder, scaffold, <llvm:PointerValue>lhs);
+            llvm:Value rhsUntagged = buildUntagFloat(builder, scaffold, <llvm:PointerValue>rhs);
+            buildCompareFloat(builder, scaffold, buildFloatCompareOp(insn.op), lhsUntagged, rhsUntagged, result);
         }
         t:UT_BOOLEAN => {
-            buildStoreBoolean(builder, scaffold,
-                              builder.iCmp(buildBooleanCompareOp(insn.op),
-                                           buildBoolean(builder, scaffold, <bir:BooleanOperand>insn.operands[0]),
-                                           buildBoolean(builder, scaffold, <bir:BooleanOperand>insn.operands[1])),
-                              insn.result);
+            llvm:Value lhsUntagged = buildUntagBoolean(builder, <llvm:PointerValue>lhs);
```

Comment:
You don't need to untag booleans to compare them.

Reply:
Fixed